### ISSUE TITLE
Confusion in tfrs.tasks.Retrieval

### ISSUE THREAD

**Issue by Akshaysharma29** [tensorflow/recommenders#637]
Hi Team, 
Thanks for creating such a wonderful library. I have a question regarding tf.keras.losses.CategoricalCrossentropy in tfrs.tasks.Retrieval.
For calculating **labels** we are using `tf.eye()` which is creating identity matrics.
So suppose we have below example:
**user | product**
  ->x1     | a
  ->x1     | b
  ->x2    | c
  
In this case, the matrix will look like the one below
 **--- a, b, c**
**x1**   [1 ,0, 0
**x1**    0, 1, 0
**x2**   0, 0, 1]

Do you think this will lead to the wrong example for the model?

**Comment by OmarMAmin** [comment]
Hi @Akshaysharma29

I guess you're right, but the library usually support the other way around, as it assumes that you have a large corpus of items, and a huge number of users, so what usually happens is that you might end up with this scenario a bit more often that what you mentioned
->x1 | a
->x2 | a
->x2 | c

In that case you can use remove_accidental_hits parameter from the model, in the Retrieval class instructor, usually you have sparse data per query (user), but the in-batch negative samples might retrieve more popular items as negatives in each batch, I guess that what you should worry about not the other way around.

Does your use case involve many interactions per user to the extent that a single user will be sampled twice in a batch?

**Comment by Akshaysharma29** [comment]
Hi @OmarMAmin 
Thanks for the answer. I will try  `remove_accidental_hits`  and yes in my case there can be multiple(maybe up to 10) instances.

**Comment by dibya-pati** [comment]
Hi @OmarMAmin, I believe the issue @Akshaysharma29  mentioned would still be valid even with "remove_accidental_hits" flag set, because, with "remove_accidental_hits" flag we would remove penalizing for the same item is the same in the batch e.g.,
user | product
->x1 | a
->x2 | a
->x2 | c
which would lead
--- a, a, c
x1 [1 ,0, 0
x1 0, 1, 0
x2 0, 0, 1]

Without the "remove_accidental_hits" flag we would consider "a" for x2 as a negative for x1, and with "remove_accidental_hits" we would assign this a high negative logit (so, loss of label, logit is extremely small). But the issue mentioned in the original question would still be valid and we would penalize for x1-b pair.

**Comment by patrickorlando** [comment]
In reality this just results in a mild regularisation effect. There are discussions on this in https://github.com/tensorflow/recommenders/issues/334#issuecomment-894873355 and in #631.

**Comment by dibya-pati** [comment]
Thanks @patrickorlando, follow up on this question. Do we know the impact of following scenarios:
1.considering a scenario where batch size > number of items, and we sample negatives, we would end up with higher contribution from popular item (occurring multiple times in batch)
2. For large batch sizes, we are essentially weighing down the impact of positive interaction and weighing up implicit negatives

**Comment by patrickorlando** [comment]
The sampled softmax is an efficient way to model massive multi-class classification. It's an approximation to the full softmax, where the positive example is 1 and every other item in the corpus is 0.

If you have upwards of 10-100K items, calculating the full softmax becomes intractable and class sampling is used.
In this case the bias caused by sampling the negative classes is corrected using the `log(Q)` sampling correction (this should be applied to avoid issues you are describing.) I would advise implementing the `candidate_sampling_probability` parameter, for which there are other issues in this repo discussing it.

If you have fewer items than the batch size, you might as well just implement the full softmax, since the approximation is not necessary.

**Comment by OmarMAmin** [comment]
The main problem with the full softmax, is that you won't have label features and you'll end up implementing an architecture like the one from earlier CG in youtube, but i guess it's worth exploring, with including item features in the context and user history somehow
<img width="888" alt="image" src="https://user-images.githubusercontent.com/7774897/221134992-7d098113-dfb6-4b86-9111-739cb1a2e728.png">
 
@patrickorlando I thought with the full softmax, we won't need to correct for the sampling, can you elaborate more why we would need that? is it because rare items aren't selected as positives that much?

I'm confused in that case how should we correct for it, if we did class imbalance correction, we're actually hurting users with popular items interaction somehow, so class imbalanace has a user-item coupling somehow

**Comment by patrickorlando** [comment]
The full softmax will include each item exactly once, therefore no correction is needed. You can still have item features and you can use sample weighting to encourage the model to pay more attention to less popular items if that makes sense for your problem.

**Comment by OmarMAmin** [comment]
Thanks for your reply, i meant for the sampled softmax you'll be having only context and user features (previous items .. etc), to incorporate items specific features, it'll be more of a ranking model rather than a retrieval one as you'll need for each user to run through the network for every single item.

The previous paper mentioned that for the retrieval style (that allows for one prediction per user with precomputed items embeddings) we have two options
1. Two Tower (where the candidate tower will include additional features per item)
2. Single Tower with the items embedding being the last softmax layer weights (as in the previous image) (with sacrificing  items specific features to have the prediction be once per user)

**Comment by patrickorlando** [comment]
@OmarMAmin, what I mean by full-softmax is that instead of using in-batch negatives, each batch computes the candidate embeddings for all items in the corpus to use as negatives. It's still compatible with a two-tower model, but it would require a modification of the `Retrieval` task.

The output logits would be `(Batch Size, Total Items)` and the labels would no longer be an identity matrix. In this case the Softmax is not being approximated, we are computing the true probability distribution. It is when this softmax becomes computationally intractable that we apply the sampled softmax for which the logQ correction is required to align the logits with this true probability distribution.

**Comment by OmarMAmin** [comment]
Got it, thanks, I didn't think about it that way, thought the only way to discard the item tower, what you're suggesting, is that for each batch, all items will be going through the item tower, and the batch will be user, context --> label, and the loss should be modified accordingly.

Great idea if the item space is manageable, thanks for sharing

**Comment by patrickorlando** [comment]
yep, exactly! you explained it much better than I did!
